### PR TITLE
BERT Fine-Tuning

### PR DESCRIPTION
**Pull request by Readrid**

### CODE REVIEW

File: text_models/abstract_model.py [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
 
 
 class AbstractModel(ABC):
     name = "abstract"
 
+    from_scratch = "TASK"
+    pretrained = "PT+TASK"
+    finetuned = "PT+DOC+TASK"
+    doc_task = "DOC+TASK"
+
```
Comment (egor-bogomolov) [inline text_models/abstract_model.py]: A good place for this is a separate enumeration

File: approaches/abstract_approach.py [hunk 2]
```diff
@@ -31,28 +31,37 @@ def evaluate_all(
...
+            AbstractModel.finetuned: finetuned_model,
         }
+        for name, model in models_dict:
+            if model is not None:
+                res_dict[name] = self.evaluate(model, topns)
 
         self.results = pd.DataFrame.from_dict(res_dict)
-        print(self.results)
+
+        if not silence:
```
Comment (egor-bogomolov) [inline approaches/abstract_approach.py]: I believe that `not silence` is [usually](https://stackoverflow.com/questions/29995249/verbose-argument-in-scikit-learn) called `verbose`

File: approaches/abstract_approach.py [hunk 3]
```diff
@@ -31,28 +31,37 @@ def evaluate_all(
...
+            print(self.results)
 
     def save_results(self, save_to_path, model_name, graph=False):
         if self.results is None:
             raise ValueError("No results to save")
 
+        os.makedirs(save_to_path, exist_ok=True)
+
         self.results.to_csv(os.path.join(save_to_path, model_name + ".csv"))
         if graph:
```
Comment (egor-bogomolov) [inline approaches/abstract_approach.py]: `graph` might sound a bit ambiguous (even though technically correct), something like `plot` or `draw` would be more clear for a reader

File: approaches/finetuning_tasks_test.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        )
+        self.results_list = [self.approach.results]
+
+        if finetuned_model is None and doc_task_model is None:
+            return
+
+        self.all_results = self.approach.results.copy()
+        self.all_results.rename(columns={AbstractModel.finetuned: f"PT+DOC({self.tasks[0]})+TASK"}, inplace=True)
+
+        for i in range(1, len(self.tasks)):
```
Comment (egor-bogomolov) [inline approaches/finetuning_tasks_test.py:46]: Why do you skip the very first element? A comment here would be really helpful to explain what's going on :slightly_smiling_face:

File: config.yml [hunk 5]
```diff
@@ -1,22 +1,40 @@
 datasets:
-  full: data/csv/IDEA2020.csv
-  train: data/csv/train2020.csv
-  test: data/csv/test2020.csv
+  full: data/csv/KT2020.csv
+  train: data/csv/train_KT.csv
+  test: data/csv/test_KT.csv
```
Comment (egor-bogomolov) [inline config.yml]: Let's create separate configs for IDEA and Kotlin and store them both (preferably in a separate directory rather than in the project root)

Comment (egor-bogomolov) [reply config.yml]: In order to avoid copying, you can store configuration in several YAML files and [merge them with omegaconf](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#omegaconf-merge)

File: text_models/bert_tasks/IREvalTrainer.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            self.device = torch.device(device)
+            self.task = task
+
+        def encode(
+            self,
+            sentences: Union[str, List[str]],
+            batch_size: int = 32,
+            show_progress_bar: bool = None,
+            output_value: str = "sentence_embedding",
+            convert_to_numpy: bool = True,
```
Comment (egor-bogomolov) [inline text_models/bert_tasks/IREvalTrainer.py]: Is it used?

File: text_models/bert_tasks/IREvalTrainer.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            self.max_len = max_len
+            self.device = torch.device(device)
+            self.task = task
+
+        def encode(
+            self,
+            sentences: Union[str, List[str]],
+            batch_size: int = 32,
+            show_progress_bar: bool = None,
+            output_value: str = "sentence_embedding",
```
Comment (egor-bogomolov) [inline text_models/bert_tasks/IREvalTrainer.py]: Is it used?

File: text_models/bert_tasks/abstract_task.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        evaluator: evaluation.InformationRetrievalEvaluator,
+        max_len: int,
+        device: str,
+        save_to_path: str,
+    ) -> models.Transformer:
+        raise NotImplementedError()
+
+    @staticmethod
+    def sections_to_sentences(docs_corpus):
+        return [" ".join(doc) for doc in flatten(docs_corpus)]
```
Comment (egor-bogomolov) [inline text_models/bert_tasks/abstract_task.py]: Is this class the best place for this func?

File: text_models/bert_tasks/masked_lm.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        self,
+        epochs=2,
+        batch_size=16,
+        eval_steps=200,
+        n_examples="all",
+        save_best_model=False,
+        mask_probability=0.15,
+        save_steps=5000,
+    ):
+        super().__init__(epochs, batch_size, eval_steps, n_examples, save_best_model, name="mlm")
```
Comment (egor-bogomolov) [inline text_models/bert_tasks/masked_lm.py]: You have names hard-coded in two places (in tasks and in module-level `__init__`). It's better to hard-code them in one place and retrieve mapping in the second one. For example, you can introduce a field in each class that stores its name. Alternatively, you can retrieve the mapping from the dict in module's `__init__`.

File: text_models/datasets/same_section.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+from text_models.datasets import BertModelDataset
+
+
+class SameSectionDataset(BertModelDataset):
+    def __init__(self, sections: List[List[str]], tokenizer, n_examples: Union[str, int] = "all", max_len: int = 512):
+        sentences_a = []
+        sentences_b = []
+
+        def randint_except(low, high, excluding):
+            return np.random.choice(np.arange(low, excluding).tolist() + np.arange(excluding + 1, high).tolist())
```
Comment (egor-bogomolov) [inline text_models/datasets/same_section.py]: You can avoid creation of lists:
```python
result = np.random.randint(low, high - 1)
return result + int(result >= excluding)
```

**Review by egor-bogomolov — APPROVED** [review]
Thanks a lot for implementing this! There are lots of missing type annotations, could you please add them?

File: text_models/abstract_model.py (hunk 1 above)
Comment (Readrid) [reply text_models/abstract_model.py]: Indeed, now it's in `text_models.abstract_model.TrainerTypes`

File: approaches/abstract_approach.py (hunk 2 above)
Comment (Readrid) [reply approaches/abstract_approach.py]: Fixed

File: approaches/abstract_approach.py (hunk 3 above)
Comment (Readrid) [reply approaches/abstract_approach.py]: Ok, it's `plot` now

File: approaches/finetuning_tasks_test.py (hunk 4 above)
Comment (Readrid) [reply approaches/finetuning_tasks_test.py:46]: We need to skip first task because we have already processed this task with the self.approach.evaluate_all() method call above. I have added comment about this to the code

File: text_models/bert_tasks/IREvalTrainer.py (hunk 6 above)
Comment (Readrid) [reply text_models/bert_tasks/IREvalTrainer.py]: Now it is used :)

File: text_models/bert_tasks/IREvalTrainer.py (hunk 7 above)
Comment (Readrid) [reply text_models/bert_tasks/IREvalTrainer.py]: And this is also being used now :) I use it the same way as `SentenceTransformer.encode()` from SBERT

File: text_models/bert_tasks/abstract_task.py (hunk 8 above)
Comment (Readrid) [reply text_models/bert_tasks/abstract_task.py]: No, now it's in `data_processing.util`

File: text_models/bert_tasks/masked_lm.py (hunk 9 above)
Comment (Readrid) [reply text_models/bert_tasks/masked_lm.py]: `name` field has been introduced to every `*Task` class

File: text_models/datasets/same_section.py (hunk 10 above)
Comment (Readrid) [reply text_models/datasets/same_section.py]: Ok, thanks, fixed

File: config.yml (hunk 5 above)
Comment (Readrid) [reply config.yml]: Now the configuration for both datasets is stored in `data/datasets_config.yml`, and when loaded, it is merged with `config.yml`

File: text_models/abstract_model.py [hunk 11]
```diff
@@ -1,22 +1,42 @@
...
 class AbstractModel(ABC):
     name = "abstract"
 
     def __init__(
-        self, vector_size=300, epochs=5, pretrained_model=None, seed=42, save_to_path="./", models_suffixes=None
+        self,
+        vector_size: int = 300,
+        epochs: int = 5,
+        pretrained_model: str = None,
```
Comment (egor-bogomolov) [inline text_models/abstract_model.py]: Typing should be `Optional[str]` if `None` is a possible value

File: text_models/abstract_model.py [hunk 12]
```diff
@@ -1,22 +1,42 @@
...
     ):
         if models_suffixes is None:
             models_suffixes = {
-                "from_scratch": ".model",
-                "pretrained": "_pretrained.model",
-                "finetuned": "_finetuned.model",
+                "from_scratch": "_task.model",
+                "pretrained": "_pt_task.model",
+                "finetuned": "_pt_doc_task.model",
+                "doc_task": "_doc_task.model",
```
Comment (egor-bogomolov) [inline text_models/abstract_model.py]: Maybe you can use `TrainTypes` defined above, it seems to be more convenient?

File: text_models/abstract_model.py [hunk 13]
```diff
@@ -26,18 +46,20 @@ def __init__(
         np.random.seed(seed)
```
Comment (egor-bogomolov) [inline text_models/abstract_model.py]: There is also [a separate seed in PyTorch](https://pytorch.org/docs/stable/generated/torch.manual_seed.html) (and possibly in other libraries). Let's define a separate function that fixes all of them somewhere in utilities and call it.

File: text_models/abstract_model.py [hunk 14]
```diff
@@ -46,25 +68,36 @@ def get_doc_embedding(self, doc):
                 size += 1
         return result if size == 0 else result / size
 
-    def get_embeddings(self, corpus):
+    def get_embeddings(self, corpus: Sentences):
         return np.array([self.get_doc_embedding(report) for report in corpus]).astype(np.float32)
```
Comment (egor-bogomolov) [inline text_models/abstract_model.py]: You can directly pass type as `np.array([...], dtype=np.float32)`

File: config.yml [hunk 15]
```diff
@@ -1,22 +1,39 @@
-datasets:
-  full: data/csv/IDEA2020.csv
-  train: data/csv/train2020.csv
-  test: data/csv/test2020.csv
+dataset: idea # you can configure datasets with configs/datasets.yml
 
 test_size: 0.1
 
+tmpdir: /mnt/disk/fedorov/tmp
```
Comment (egor-bogomolov) [inline config.yml]: Please, avoid absolute paths in code and configuration files as they will break on a new machine. :pray: 

You can get the temporary directory programmatically in Python, consider doing so

File: data_processing/docs_preprocessor.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+from tika import parser
+from docutils.core import publish_string
+
+
+from util import preprocess, Sentences, Sections
+
+
+class DocsPreprocessor:
+    def __init__(self, files_path: str, extensions: List[str]):
+        self.extensions = ["." + extension for extension in extensions]
```
Comment (egor-bogomolov) [inline data_processing/docs_preprocessor.py]: It would be nice to add dots only if the string does not have a dot already (so both `ext` and `.ext` would be supported)

File: data_processing/util.py [hunk 17]
```diff
@@ -1,40 +1,53 @@
...
 from nltk.corpus import stopwords
-from nltk import FreqDist
-
-from json import JSONEncoder
-from omegaconf import OmegaConf
+from omegaconf import OmegaConf, DictConfig, ListConfig
 
 CONFIG_PATH = "config.yml"
+Sections = List[List[List[str]]]
+Sentences = List[List[str]]
```
Comment (egor-bogomolov) [inline data_processing/util.py]: Do you really need types for plural rather than for single entities? The whole system can be reformulated as follows:
```python
Sentence = List[str]
Sentences = List[Sentence]
Section = Sentences
Sections = List[Section]
```

But I would say that it is more logical to keep only `Sentence` and `Section` and use `List[Section]` when needed.

File: data_processing/util.py [hunk 18]
```diff
@@ -1,40 +1,53 @@
...
+def get_corpus(data: pd.DataFrame, sentences: bool = False) -> Union[Sentences, Sections]:
     corpus = []
     for str_list in data["description"].tolist():
         corpus.append(ast.literal_eval(str_list))
     return flatten(corpus) if sentences else list(map(flatten, corpus))
 
 
-def flatten(t):
-    return [item for sublist in t for item in sublist]
+def flatten(matrix):
```
Comment (egor-bogomolov) [inline data_processing/util.py]: What is the type of `matrix`? If it is `np.ndarray`, then there is `matrix.flatten()` method. Thus, you should specify the type to avoid misunderstanding.

File: data_processing/util.py (hunk 17 above)
Comment (egor-bogomolov) [reply data_processing/util.py]: As you need to refer to multiple sections frequently, you can keep `Sections` also or give some other name like `Corpus`

File: approaches/finetuning_tasks_test.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+                self.all_results[f"DOC({self.tasks[i]})+TASK"] = task_doc_test_res
+                self.results_list.append(res_copy)
+
+        if verbose:
+            print(self.all_results)
+
+    def save_results(self, save_to_path: str, model_name: str, plot: bool = False):
+        for i, result in enumerate(self.results_list):
+            self.results = result
+            super().save_results(save_to_path, model_name + "_" + self.tasks[int(i / 2)], plot)
```
Comment (egor-bogomolov) [inline approaches/finetuning_tasks_test.py]: Where does division by 2 come from?

File: text_models/bert_tasks/abstract_task.py [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+        self.batch_size = batch_size
+        self.n_examples = n_examples
+        self.eval_steps = eval_steps
+        self.save_best_model = save_best_model
+
+    @abstractmethod
+    def finetune_on_docs(
+        self,
+        pretrained_model: str,
+        docs_corpus: List[List[List[str]]],  # list of list(sections) of list(sentences) of tokens(words)
```
Comment (egor-bogomolov) [inline text_models/bert_tasks/abstract_task.py]: You can use the introduced types here

File: text_models/bert_tasks/masked_lm.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        save_steps: int = 5000,
+    ):
+        super().__init__(epochs, batch_size, eval_steps, n_examples, save_best_model)
+        self.mask_probability = mask_probability
+        self.save_steps = save_steps
+
+    def finetune_on_docs(
+        self,
+        pretrained_model: str,
+        docs_corpus: List[List[List[str]]],  # list of list(sections) of list(sentences) of tokens(words)
```
Comment (egor-bogomolov) [inline text_models/bert_tasks/masked_lm.py]: You can use the introduced typealiases.

File: text_models/datasets/next_sentence.py [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+        label = []
+        lngth = len(corpus) - 1
+        for i in range(lngth - forget_const):
+            if np.random.rand() >= 0.5:
+                sentence_a.append(corpus[i])
+                sentence_b.append(corpus[i + 1])
+                label.append(0)
+            else:
+                sentence_a.append(corpus[i])
+                sentence_b.append(corpus[i + np.random.randint(forget_const, lngth - i)])
```
Comment (egor-bogomolov) [inline text_models/datasets/next_sentence.py]: This way you will generate significantly more examples with examples at the end of the corpus. Consider changing the generation function to avoid this bias.

**Review by egor-bogomolov — APPROVED** [review]
Great work! There are some minor comments left to fix and after that you can merge.

File: text_models/abstract_model.py (hunk 11 above)
Comment (Readrid) [reply text_models/abstract_model.py]: Fixed

File: text_models/abstract_model.py (hunk 12 above)
Comment (Readrid) [reply text_models/abstract_model.py]: I got rid of this configuration. Now I use `TrainTypes` everywhere

File: text_models/abstract_model.py (hunk 14 above)
Comment (Readrid) [reply text_models/abstract_model.py]: Fixed

File: config.yml (hunk 15 above)
Comment (Readrid) [reply config.yml]: It was changed null. If tmpdir is null default temp directory will be used.

File: data_processing/docs_preprocessor.py (hunk 16 above)
Comment (Readrid) [reply data_processing/docs_preprocessor.py]: Fixed

File: data_processing/util.py (hunk 17 above)
Comment (Readrid) [reply data_processing/util.py]: OK, now I use `Sentence`, `Section` and `Corpus`

File: data_processing/util.py (hunk 18 above)
Comment (Readrid) [reply data_processing/util.py]: It is for python lists. I added corresponding typing.

File: approaches/finetuning_tasks_test.py (hunk 19 above)
Comment (Readrid) [reply approaches/finetuning_tasks_test.py]: It was a little bug :) Now it's fixed, thank you!

File: text_models/bert_tasks/abstract_task.py (hunk 20 above)
Comment (Readrid) [reply text_models/bert_tasks/abstract_task.py]: Fixed

File: text_models/bert_tasks/masked_lm.py (hunk 21 above)
Comment (Readrid) [reply text_models/bert_tasks/masked_lm.py]: Fixed

File: text_models/datasets/next_sentence.py (hunk 22 above)
Comment (Readrid) [reply text_models/datasets/next_sentence.py]: Now I choose negative example as random sentence in range `0..len(corpus)` excluding `[i - forget_const, i + forget_const]`

File: text_models/abstract_model.py (hunk 13 above)
Comment (Readrid) [reply text_models/abstract_model.py]: OK, `fix_random_seed(seed: int)` method has been added to `data_processing/util.py`
